node/core/candidate-validation/src/lib.rs

```diff
@@ -0,0 +1,913 @@
+	validation_pool: Option<ValidationPool>,
+	descriptor: CandidateDescriptor,
+	pov: Arc<PoV>,
+	spawn: impl SpawnNamed + 'static,
+) -> SubsystemResult<Result<ValidationResult, ValidationFailed>> {
+	// The candidate descriptor has a `validation_data_hash` which corresponds to
+	// one of up to two possible values that we can derive from the state of the
+	// relay-parent. We can fetch these values by getting the `global_validation_data`,
+	// and both `local_validation_data` based on the different `OccupiedCoreAssumption`s.
+	let global_validation_data = {
+		let (tx, rx) = oneshot::channel();
+		let res = runtime_api_request(
+			ctx,
+			descriptor.relay_parent,
+			RuntimeApiRequest::GlobalValidationData(tx),
+			rx,
+		).await?;
+
+		match res {
+			Ok(g) => g,
+			Err(e) => {
+				log::warn!(
+					target: "candidate_validation",
+					"Error making runtime API request: {:?}",
+					e,
+				);
+
+				return Ok(Err(ValidationFailed));
+			}
+		}
+	};
+
+	match check_assumption_validation_data(
+		ctx,
+		&descriptor,
+		&global_validation_data,
+		OccupiedCoreAssumption::Included,
+	).await? {
+		AssumptionCheckOutcome::Matches(omitted_validation, validation_code) => {
+			return spawn_validate_exhaustive(
+				ctx,
+				validation_pool,
+				omitted_validation,
+				validation_code,
+				descriptor,
+				pov,
+				spawn,
+			).await;
+		}
+		AssumptionCheckOutcome::DoesNotMatch => {},
+		AssumptionCheckOutcome::BadRequest => return Ok(Err(ValidationFailed)),
+	}
+
+	match check_assumption_validation_data(
+		ctx,
+		&descriptor,
+		&global_validation_data,
+		OccupiedCoreAssumption::TimedOut,
+	).await? {
+		AssumptionCheckOutcome::Matches(omitted_validation, validation_code) => {
+			return spawn_validate_exhaustive(
+				ctx,
+				validation_pool,
+				omitted_validation,
+				validation_code,
+				descriptor,
+				pov,
+				spawn,
+			).await;
+		}
+		AssumptionCheckOutcome::DoesNotMatch => {},
+		AssumptionCheckOutcome::BadRequest => return Ok(Err(ValidationFailed)),
+	}
+
+	// If neither the assumption of the occupied core having the para included or the assumption
+	// of the occupied core timing out are valid, then the validation_data_hash in the descriptor
+	// is not based on the relay parent and is thus invalid.
+	Ok(Ok(ValidationResult::Invalid))
+}
+
+async fn spawn_validate_exhaustive(
+	ctx: &mut impl SubsystemContext<Message = CandidateValidationMessage>,
+	validation_pool: Option<ValidationPool>,
+	omitted_validation: OmittedValidationData,
+	validation_code: ValidationCode,
+	descriptor: CandidateDescriptor,
+	pov: Arc<PoV>,
+	spawn: impl SpawnNamed + 'static,
+) -> SubsystemResult<Result<ValidationResult, ValidationFailed>> {
+	let (tx, rx) = oneshot::channel();
+	let fut = async move {
+		let res = validate_candidate_exhaustive::<RealValidationBackend, _>(
+			validation_pool,
+			omitted_validation,
+			validation_code,
+			descriptor,
+			pov,
+			spawn,
+		);
+
+		let _ = tx.send(res);
+	};
```

Comment:
According to the guide, this should have a timeout, or are we expecting the runtime to take care of that?

Reply:
The `wasm_executor` crate below has the time-out functionality